File: java-to-python-practice/best-time-to-buy-and-sell-stock/best_time_to_buy_and_sell_stock.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        max_achievable_profit = 0
        min_price_so_far = prices[0]
        
        for i in range(1, len(prices)):
            max_profit_if_sell_at_cur_price = prices[i] - min_price_so_far
            
            min_price_so_far = min(prices[i], min_price_so_far)
            max_achievable_profit = max(max_achievable_profit, max_profit_if_sell_at_cur_price)
            
        return max_achievable_profit
        

File: java-to-python-practice/best-time-to-buy-and-sell-stock/bestTimeToBuyAndSellStock.java
class Solution {
    public int maxProfit(int[] prices) {
        int maxAchievableProfit = 0;
        int minPriceSoFar = prices[0];
        
        for (int i = 1; i < prices.length; i++) {
            int maxProfitIfSellAtCurPrice = prices[i] - minPriceSoFar;
            
            minPriceSoFar = Math.min(prices[i], minPriceSoFar);
            maxAchievableProfit = Math.max(maxAchievableProfit, maxProfitIfSellAtCurPrice);
        }

        return maxAchievableProfit;
    }
}